### PR TITLE
Added basic functionality for creating new magical spells

### PR DESCRIPTION
**Pull request by Gregorvich**
Hey Jacob, I created a file that adds the functionality for our shopping cart to cast magic in case we want our shopping cart to be able to do that. It is basically just a class definition for Sorceries, and then a couple functions that were more meant for practicing some python concepts and demonstrating what can be done with a spell book (a list of sorceries). We don't have to keep this feature; I mostly created the program just to have something that we can merge into the master branch.

Also, I was looking back at your requirements for solving the initial issue, and I noticed that I hadn't created a file for our source code that is separate from the source file of the repository, but I figured you could explain some of those concepts in more detail as we proceed. These would include things such as properly organizing our project, what is the standard naming convention for branches, etc.

Let me know if there is something additional that you want me to do before we are able to merge these changes to the master branch

### CODE REVIEW

File: Sample_Program1.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+class Sorcery:
+    'Common base class for magic spells'
+    sorceryCount = 0
```
Comment (Jacob-Gregor) [inline Sample_Program1.py]: `sorceryCount` should be in __init__ ()

Along with that, how does that variable work? Each instance of Sorcery() being created on lines 25, 26, 27 each have it's own count?

File: Sample_Program1.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    print("Checking to see if all spells in the spell book are valid...")
+    for i in range(len(spellBook)):
+        if (spellBook[i].getName() == hippoSpellName):
+            hippoFlag = True
+            forbiddenSpellLocation = i
+
+    spellBook[0].displayCount()
+    if (hippoFlag):
+        print("The hippopotamus spell is too powerful. It")
```
Comment (Jacob-Gregor) [inline Sample_Program1.py]: You will want to add a space after "It"

File: Sample_Program1.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        print("should not be added to the spellBook!")
```
Comment (Jacob-Gregor) [inline Sample_Program1.py]: You will want to add a space after "spellBook!"

File: Sample_Program1.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        print("Here's some info on it anyway:")
```
Comment (Jacob-Gregor) [inline Sample_Program1.py]: You will want to add a space after ":"

**Review by Jacob-Gregor — CHANGES_REQUESTED** [review]
You definitely went beyond what said you would deliver. Kudos. We will proceed with what you got. And what I mean by that is if this is the code you want to merge, then it has to be good and polished.

For the body of the PR message, you will only want to add information that is relevant to the code. It makes it easier for reviewers, like me, to understand the purpose of the pull request, what goals are the pull request trying to achieve, and how the PR accomplishes it.  Do not add any fluff in the PR body. If you have extra comments put it in the comments

 Along with my comments inline (in your code), I have the following that needs to be addressed.

- Rename `Sample_Program1.py` to a name that better represents the work that is being done in the file. In the issue I asked for a main.py file.
- Create a directory specifically for scripts
- Part of the instructions is for me to execute this file. How am I supposed to do that based on what you coded here? I do not see instructions. l cannot run `python Sample_Program1.py`, since it wouldn't do anything. Do you want me to only execute a specific function in the file?

I have other questions and comments, but I want to run your file before I make them, just to make sure if my comments are valid or not.

File: Sample_Program1.py (hunk 4 above)
Comment (Gregorvich) [reply Sample_Program1.py]: Rather than renaming the Sample_Program1.py file to Main.py, I renamed it to Sorcery_Functions.py. I'm not sure if that's what you suggested, but there you have it. I then created a Main.py to use to test the Sorcery_Functions.py file. The Main.py file is basically a script meant for testing right now, but we can change the functionality of it later if you want. In the Main.py file I already imported the Sorcery_Functions module, and after the import I added some test cases so that testing it is a bit more streamlined. Therefore, to execute the tests you simply have to import the Main.py file using the Python command line, and then run the function testSorceries() in the Main module. So I guess that Main.py is really more of a module than a script since it has a function definition, and you have to import the module and run a function from it, but I guess it's whatever you want to call it.

Let me know if you have issues testing the Sorcery_Functions module.

Comment (Gregorvich) [reply Sample_Program1.py]: One more thing I forgot to mention. I know you said to add spaces to the end of my print() statements, but I'm not entirely sure what the point is because, for my interpreter anyway, the print() function automatically adds a new line after printing the text inputted into it. Was there a specific reason to add spaces at the end of the print() statements? I can definitely see it being an issue if I were chaining together output text onto a single line with multiple print() statements, but I'm not really sure if you can even do that in Python.

File: Scripts/Main.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+# The below if statement is added in case we want to input
+# arguments in the command line for testing the Sorcery function.
+# So it's mainly added for convenience, since an alternative
+# would be to manually edit this script. When command line
+# arguments are used, the Main function expects the first
+# argument to be a boolean, although if this is not the case
+# it will type cast the argument to type "bool"
+# if (__name__ == "__main__"):
+#     import sys
```
Comment (Jacob-Gregor) [inline Scripts/Main.py]: Import should never be in an if statement in my opinion. If you or another person has to do that, than it might be a big clue that the code needs to be reorganized in some way.

File: Scripts/Sorcery_Functions.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+# will print a list of the spells in the spell book. It
+# will not print the list if the size of the spell book
+# is 0. Additionally, the type of output printed will vary
+# depending on whether the spell book contains forbidden
+# spells
+def printSpellBook(spellBook):
+    hippoSpellName = "Summon Hippopotamus"
+    hippoFlag = False
+    forbiddenSpellFlag = False
+    hippoSpellLocation = -1
```
Comment (Jacob-Gregor) [inline Scripts/Sorcery_Functions.py]: hippoSpellLocation shoud probably be set to Null instead of -1 since -1 is a location in a list.

File: Scripts/Sorcery_Functions.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        print("should not be added to the spellBook! ")
+        print("Here's some info on it anyway: ")
+        spellBook[hippoSpellLocation].displayMagicAttributes()
+
+    elif(forbiddenSpellFlag):
+        print("Caution: At least one forbidden spell was found.")
+        print("Here is a complete list of the forbidden spells")
+        print("found in the spell book:")
+        for i in range(len(spellBook)):
+            if (spellBook[i].getForbiddenStatus()):
```
Comment (Jacob-Gregor) [inline Scripts/Sorcery_Functions.py]: Something to make note of, you can write this for loop like this:

` for spell in spellbook:`
`    if spell.getForbidenStatus():`
`      spell.displayMagicAttributes()`

Couldn't format it cleanly, but give it a try.  I can break down why this works if you want.

**Review by Jacob-Gregor — CHANGES_REQUESTED** [review]
You should remove Hello_World.py. It isn't relevant to your code.

I also have comments inline for you to look at. I still have to test this code, so once I do that I may have more comments.

File: Scripts/Main.py (hunk 5 above)
Comment (Gregorvich) [reply Scripts/Main.py]: I left this part of the code untouched for now since it is commented out, and, once you get done with testing the SorceryFunctions.py file, you can let me know if you just want me to remove that section. I guess the reason the import statement is in the if statement is because, if the arguments for the functions we're wanting to test aren't put straight into the command line, then the sys module isn't used, but that functionality isn't really needed anyway because I already streamlined the testing of functions in the SorceryFunctions.py file by creating a testSorceries() function in Main.py. I guess common practice is to import the modules that we need at the beginning of scripts, so I can see why this is an eyesore.

File: Scripts/Sorcery_Functions.py (hunk 6 above)
Comment (Gregorvich) [reply Scripts/Sorcery_Functions.py]: Ah, I forgot that Python is flexible in that it allows negative indices. I went ahead and changed the initialization to the singleton None.

**Comment by Gregorvich** [comment]
Just so that you do not miss it, I made two commits since you last checked the code. After pushing to Github, I realized there was another bug fix I wanted to make quick. So things should be working properly now, but let me know if you have any issues running the code.

File: Scripts/Main.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+#     includeForbiddenSpells = bool(sys.argv[1])
+#     createSpells = Sorcery_Functions.createSpells
+#     printSpellBook = Sorcery_Functions.printSpellBook
```
Comment (Jacob-Gregor) [inline Scripts/Main.py]: Lines 13 to 16 are over indented.

I am also running an error on line 14 (Line 15 for me since I added a blank line locally)

Traceback (most recent call last):
  File "Scripts/Main.py", line 15, in <module>
    includeForbiddenSpells = bool(sys.argv[1])
IndexError: list index out of range

**Review by Jacob-Gregor — CHANGES_REQUESTED** [review]
Tried testing, ran into issues. Comments inline.

**Comment by Jacob-Gregor** [comment]
Do you have any questions? Haven't heard anything in a while.

**Comment by Gregorvich** [comment]
I made two minor changes. First, I deleted the commented out code that was used for command line testing. I did mention earlier that we could just ignore it for now, but it's gone now so that will make things simpler. The second thing I did is that I made it somewhat more streamlined to test the code. What I did is that at the end of the Main script, there is a function call to the testSorceries() function, so now the function will run once Main is imported; There is no need to then run Main.testSorceries(). I guess the main effect this has is that now, from the command line and NOT the Python bash, you can run "Python Main.py" in order to test the Sorcery_Functions module.

So, to recap, here are the two methods for testing the Sorcery_Functions module:

1. Once in the Scripts file, run an instance of Python Bash, then import the Main module. Once that is done, you should be able to run functions in the Main module, such as the testSorceries() function.
2. Once in the Scripts file, simply run "Python Main.py" from the command line.

Let me know if you have any issues running the code and/or there are bugs that need to be addressed.

**Comment by Jacob-Gregor** [comment]
Cool. I'll test it tonight.

File: Scripts/Main.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    print("forbidden spell")
+    overpoweredSpellBook = createSpells(False, True)
+    printSpellBook(overpoweredSpellBook)
+    print()
+
+    print("Testing a spell book that contains some forbidden spells")
+    print("and the hippopotamus spell")
+    reallyOverpoweredSpellBook = createSpells(True, True)
+    printSpellBook(reallyOverpoweredSpellBook)
+    print()
```
Comment (Jacob-Gregor) [inline Scripts/Main.py:34]: This print() doesn't do anything

File: Scripts/Main.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    evilSpellBook = createSpells(True, False)
+    printSpellBook(evilSpellBook)
+    print()
+    
+    print("Testing a spell book that contains the dreaded and")
+    print("spooky hippopotamus spell, but not any other kind of")
+    print("forbidden spell")
+    overpoweredSpellBook = createSpells(False, True)
+    printSpellBook(overpoweredSpellBook)
+    print()
```
Comment (Jacob-Gregor) [inline Scripts/Main.py:28]: This print doesn't do anything

File: Scripts/Main.py [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    print("Forbidden Spells")
+    normalSpellBook = createSpells(False, False)
+    printSpellBook(normalSpellBook)
+    print()
+    
+    print("Testing a spell book that contains forbidden spells")
+    print("but NOT the infamous hippopotamus spell")
+    evilSpellBook = createSpells(True, False)
+    printSpellBook(evilSpellBook)
+    print()
```
Comment (Jacob-Gregor) [inline Scripts/Main.py:21]: This print() doesn't do anything

**Review by Jacob-Gregor — APPROVED** [review]
I have a few nits, but I think this now addresses all the points issue 1 wanted. 

Approved. I will merge, then create another issue with the next task.

**Comment by Jacob-Gregor** [comment]
Important thing to note too is that once you merge a branch into master, you should always delete the merge branch from GitHub.
